### ISSUE TITLE
io.realm.exceptions.RealmMigrationNeededException: Field count does not match - expected x but was y (0.85.1)

### ISSUE THREAD

**Issue by ashemah** [realm/realm-java#1825]
Hi all, since upgrading to the latest version of realm (0.85.1) my migrations are broken and it is driving me crazy.

    io.realm.exceptions.RealmMigrationNeededException: Field count does not match - expected x but was y

I keep getting the exception above even though I have a realm migration specified in the config that covers the migration. I have tried every combination of manually handling the RealmMigrationNeededException and migrating the realm or setting a migration in the config and I still get either the error above or a migration needed error or similar.

Also, I have added @Required to all of my String fields and I was originally running on 0.80.1 if that helps.

Does anyone have an idea why this could be happening?

My migration stuff in my Application.onCreate()

```java
        RealmConfiguration config0 = new RealmConfiguration.Builder(this)
                .schemaVersion(4)
                .migration(new Migration())
                .build();

        try {
            Realm realm1 = Realm.getInstance(config0);
        }
        catch (RealmMigrationNeededException e) {
            Realm.migrateRealm(config0, new Migration());
        }

        Realm.setDefaultConfiguration(config0);

      // If I call Realm realm1 = Realm.getDefaultInstance() here it will still explode.
```

My migration

```java
package com.foo.bar;

import com.foo.bar.data.Answer;
import com.foo.bar.data.AnswerSet;
import com.foo.bar.data.Survey;
import com.foo.bar.data.User;

import io.realm.Realm;
import io.realm.RealmMigration;
import io.realm.internal.ColumnType;
import io.realm.internal.Table;

public class Migration implements RealmMigration {

    @Override
    public long execute(Realm realm, long version) {

        if (version == 0) {
            Table locationTable = realm.getTable(User.class);
            locationTable.addColumn(ColumnType.STRING, "pushToken");

            Table answerSetTable = realm.getTable(AnswerSet.class);
            answerSetTable.addColumn(ColumnType.INTEGER, "firstReminderRequestCode");
            answerSetTable.addColumn(ColumnType.INTEGER, "secondReminderRequestCode");
            answerSetTable.addColumn(ColumnType.INTEGER, "firstReminderTimestamp");
            answerSetTable.addColumn(ColumnType.INTEGER, "secondReminderTimestamp");

            version++;
        }

        if (version == 1) {
            Table surveyTable = realm.getTable(Survey.class);
            surveyTable.addColumn(ColumnType.BOOLEAN, "scheduleAllowMonday");
            surveyTable.addColumn(ColumnType.BOOLEAN, "scheduleAllowTuesday");
            surveyTable.addColumn(ColumnType.BOOLEAN, "scheduleAllowWednesday");
            surveyTable.addColumn(ColumnType.BOOLEAN, "scheduleAllowThursday");
            surveyTable.addColumn(ColumnType.BOOLEAN, "scheduleAllowFriday");
            surveyTable.addColumn(ColumnType.BOOLEAN, "scheduleAllowSaturday");
            surveyTable.addColumn(ColumnType.BOOLEAN, "scheduleAllowSunday");

            version++;
        }

        if (version == 2) {
            Table answerTable = realm.getTable(Answer.class);
            answerTable.addColumn(ColumnType.INTEGER, "reactionTimeMs");

            version++;
        }

        if (version == 3) {
           // blank just for testing purposes - I didn't actually change the model and this error was still happening when this migration was not here.
            version++;
        }

        return version;
    }
}
```

**Comment by cmelchior** [comment]
What are the x and y values in your error message? That might give us a better idea what is happening. 
```
io.realm.exceptions.RealmMigrationNeededException: Field count does not match - expected x but was y
```

**Comment by ashemah** [comment]
Hi, so when i run it now the error is:

    RealmMigrationNeededException: Field count does not match - expected 8 but was 9

I can see only 2 migrations that added 1 field. The migrations to User and Answer. And answer is the only model with 8 fields. 

I'm not really sure what the error message is referring to unfortunately.

Here is my Answer model:

```java
public class Answer extends RealmObject {

    public AnswerSet getSet() {
        return set;
    }

    public void setSet(AnswerSet set) {
        this.set = set;
    }

    public int getDbSurveyId() {
        return dbSurveyId;
    }

    public void setDbSurveyId(int dbSurveyId) {
        this.dbSurveyId = dbSurveyId;
    }

    public int getDbQuestionSetId() {
        return dbQuestionSetId;
    }

    public void setDbQuestionSetId(int dbQuestionSetId) {
        this.dbQuestionSetId = dbQuestionSetId;
    }

    public int getDbQuestionId() {
        return dbQuestionId;
    }

    public void setDbQuestionId(int dbQuestionId) {
        this.dbQuestionId = dbQuestionId;
    }

    public String getAnswerValue() {
        return answerValue;
    }

    public void setAnswerValue(String answerValue) {
        this.answerValue = answerValue;
    }

    public long getDisplayedTimestamp() {
        return displayedTimestamp;
    }

    public void setDisplayedTimestamp(long displayedTimestamp) {
        this.displayedTimestamp = displayedTimestamp;
    }

    public long getAnsweredTimestamp() {
        return answeredTimestamp;
    }

    public void setAnsweredTimestamp(long answeredTimestamp) {
        this.answeredTimestamp = answeredTimestamp;
    }

    public long getReactionTimeMs() {
        return reactionTimeMs;
    }

    public void setReactionTimeMs(long reactionTimeMs) {
        this.reactionTimeMs = reactionTimeMs;
    }

    private AnswerSet set;
    private int dbSurveyId;
    private int dbQuestionSetId;
    private int dbQuestionId;

    @Required
    private String answerValue;

    private long displayedTimestamp;
    private long answeredTimestamp;

    private long reactionTimeMs;
}
```

**Comment by ashemah** [comment]
And the entire stacktrace:

```
E/AndroidRuntime(13174): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2198)
E/AndroidRuntime(13174): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2257)
E/AndroidRuntime(13174): 	at android.app.ActivityThread.access$800(ActivityThread.java:139)
E/AndroidRuntime(13174): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1210)
E/AndroidRuntime(13174): 	at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime(13174): 	at android.os.Looper.loop(Looper.java:136)
E/AndroidRuntime(13174): 	at android.app.ActivityThread.main(ActivityThread.java:5086)
E/AndroidRuntime(13174): 	at java.lang.reflect.Method.invokeNative(Native Method)
E/AndroidRuntime(13174): 	at java.lang.reflect.Method.invoke(Method.java:515)
E/AndroidRuntime(13174): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
E/AndroidRuntime(13174): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
E/AndroidRuntime(13174): 	at dalvik.system.NativeStart.main(Native Method)
E/AndroidRuntime(13174): Caused by: io.realm.exceptions.RealmMigrationNeededException: Field count does not match - expected 8 but was 9
E/AndroidRuntime(13174): 	at io.realm.AnswerRealmProxy.validateTable(AnswerRealmProxy.java:234)
E/AndroidRuntime(13174): 	at io.realm.DefaultRealmModuleMediator.validateTable(DefaultRealmModuleMediator.java:80)
E/AndroidRuntime(13174): 	at io.realm.Realm.initializeRealm(Realm.java:339)
E/AndroidRuntime(13174): 	at io.realm.Realm.createAndValidate(Realm.java:307)
E/AndroidRuntime(13174): 	at io.realm.Realm.create(Realm.java:250)
E/AndroidRuntime(13174): 	at io.realm.Realm.getDefaultInstance(Realm.java:198)
E/AndroidRuntime(13174): 	at com.foo.bar.activity.LoginActivity.onCreate(LoginActivity.java:82)
E/AndroidRuntime(13174): 	at android.app.Activity.performCreate(Activity.java:5248)
E/AndroidRuntime(13174): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1110)
E/AndroidRuntime(13174): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2162)
E/AndroidRuntime(13174): 	... 11 more
```

**Comment by cmelchior** [comment]
You can try to iterate over the class to see which fields are there. That should hopefully give an idea which field might be the extra.

```
Table t = realm.getTable(Answer.class);
for (int i = 0; i < t.getColumnCount(); i++) {
  Log.i("Migration", t.getColumnName(i);
}
```

**Comment by ashemah** [comment]
Ok thanks, I'll give it a shot.

**Comment by ashemah** [comment]
Ok the plot thickens... (or thinnens depending on your perspective). Hopefully this helps someone else.

I think Realm was complaining because it was auto-migrating the fields (adding them) and so I didn't need to also add them in my Migration.

I can't tell if this auto-migration only just started happening or the doco is still out of date or something else. Either way thanks heaps for your quick response and help!

**Comment by kneth** [comment]
If you add fields to your model class, you must add a migration step which also adds the fields to the Realm file. That way the model classes and Realm file will match.

**Comment by kneth** [comment]
@ashemah Did you get it to work by adding a migration step?

**Comment by ashemah** [comment]
Hi Kenneth, sorry for not following up on this. I was going to close it after I had confirmed what I was seeing.

Basically, when running 0.85.1 I found that I was getting duplicate migrations if I manually added the fields myself in the Migration class.

Are you saying that now you just need to bump the migration version but not actually add any fields (if it is a simple added-new-fields migration)? That makes sense to me. The only thing that doesn't make sense is that I didn't run into this issue previously.

In one of my apps I my Realm was at version 2 and was migrating successfully when people upgraded (I think). My migration class was adding the fields manually to the realm as I thought that was what I had to do (this was done a while ago on version 0.80.1 I think).

Then I upgraded Realm in the app to 0.85.1 and it started exploding when users tried to upgrade to the new version of the app. After looking into it I worked out that I just need to remove the manual migrations entirely and let Realm automigrate people to the new schema. Does this sound right or am I crazy? :)
